flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedPartitionRequestClientHandler.java

```diff
@@ -433,8 +459,8 @@ public void operationComplete(ChannelFuture future) throws Exception {
 
         @Override
         public Object buildMessage() {
-            return new AddCredit(
-                    inputChannel.getAndResetUnannouncedCredit(), inputChannel.getInputChannelId());
+            int credits = inputChannel.getAndResetUnannouncedCredit();
+            return credits > 0 ? new AddCredit(credits, inputChannel.getInputChannelId()) : null;
```

Comment:
Why is this an issue? Is this an independent optimisation or is it relate with the other parts of the PR?

Reply:
Secondly, return type of the `buildMessage()` is not annotated `@Nullable`, so if we need this code for correctness we need to add this annotation (but it would be better to avoid `null` here)

Reply:
It is relevant to this PR, it is because we add the reset logic in ```RemoteInputChannel#resumeConsumption``` method.
```
        if (initialCredit == 0) {
            unannouncedCredit.set(0);
        }
```

Reply:
I will add a @Nullable annotation.